### ISSUE TITLE
Validating the CTF Estimator

### ISSUE THREAD

**Issue by j-c-c** [ComputationalCryoEM/ASPIRE-Python#742]
Current testing for `estimate_ctf` uses hardcoded values that have not been validated. The individual formulas also need to be validated. One possible edit mentioned here: https://github.com/ComputationalCryoEM/ASPIRE-Python/pull/724#issuecomment-1307598662

**Comment by chris-langfield** [comment]
For reference with `CTFFIND4` benchmarking:

This program outputs defocus values in **angstroms**, and defocus angle in **degrees**.

See Rohou A, Grigorieff N. CTFFIND4: Fast and accurate defocus estimation from electron micrographs. J Struct Biol. 2015 https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6760662/pdf/nihms-1021825.pdf

Specifically Fig. 4 on Page 15

**Comment by chris-langfield** [comment]
RELION 4.0 and 3.1 both use this program under the hood

**Comment by garrettwrong** [comment]
Chris, before you start more new things we should probably finish this project started a few weeks ago.  IIRC the initial results did not validate.

**Comment by garrettwrong** [comment]
This is going to come in two parts.  

First, I am going to revert the ctf estimator code to last release and fix the unit test values based on the CTFFIND numbers (understanding that they are off and that the angle might be junk).  Will aim for 10.1 on these parts.

Second we will continue to validate and maybe ask for some help from the original author regarding the angle. This will be in 10.2.

**Comment by chris-langfield** [comment]
Worth noting:

Going back to the original code, the `CTFEstimator` saves its `pixel_size` parameter as `_rlnDetectorPixelSize`. In the code, the `pixel_size` is assumed to be in Angstroms, however, `_rlnDetectorPixelSize` is properly an attribute of *movies*, not micrographs. In addition, its units are in *micrometers*.

To me it looks like this is just a naming mistake, and the save column should be changed to `_rlnMicrographPixelSize`, which is in Angstroms.

This connects to #806 because `Source`s will need to know exactly what they are looking for when reading pixel sizes from STAR files. 

From `relion_refine --print_metadata_labels`:

```
rlnDetectorPixelSize (double) : Pixel size of the detector (in micrometers)

```

Patch in: #817

**Comment by garrettwrong** [comment]
Thanks Chris.  Do you by chance know what the default (ASPIRE) behavior would be if that value is not in the starfile?

**Comment by chris-langfield** [comment]
So currently when importing CTF information from ASPIRE files, `CoordinateSource` (erroneously) *requires* `_rlnDetectorPixelSize`. 

It may be a question whether this should be required or whether a pixel size parameter should be passable to `CoordinateSource` as a fallback.

**Comment by garrettwrong** [comment]
Fair question.  I think it is probably okay for now to leave it required.  If it is missing maybe we can print a message about adding it to starfile.

The problem with the arg is that we'll have to deal with when it is in both places... and I see that potentially being trouble for us and for users.

**Comment by chris-langfield** [comment]
Agreed, I'll note that we want to add an error message when this change is applied to `CoordinateSource`

**Comment by garrettwrong** [comment]
I think the forward model is reasonably validated.  Some notes.

There is a discrepancy between the behavior of MATLAB code and Python code for even and odd resolutions (😂 ).  For odd resolutions we achieve nice parity even though we use a different formula.  I also confirmed porting the MATLAB function to Python yields a good correspondence. I have added a comment with a variant of the legacy formula for posterity.

For the radially symmetric CTF, CTFFIND4 appears to estimate our `RadialCTFFilter` defocus parameters pretty well.  It is performing much better than our internal `CtfEstimator`.  I have included the results as part of a CTF gallery tutorial PR.

The performance for asymmetric CTFs is generally not well.  I have some confidence we are modeling them okay, but the estimates are questionable enough to avoid using IMO.  I intentionally avoided this in the gallery, other than a note that it is in progress, for diplomatic reasons.

The remaining task is to attempt validating the other direction, that is, taking some data with CTF corruptions modelled from another tool like Relion, and seeing if our Source and CTFFilter are populated correctly and can apply a reasonable phase flip.

**Comment by garrettwrong** [comment]
Elaborating on the above remark about general ctf estimation.  Here are the results of estimating two outputs of our CTF model.  ( Using a single 512,512 image with a very small amount of white noise).

CTFFIND4 agrees with our forward model. Tried various invocations, similar results; depends how much time you want to wait. Overall, happy with this part. 

```
# Radial, Modeled defocus 10000
Estimated defocus values        : 9982.50 , 9982.50 Angstroms
```

```
# Modeled defocus u, v: 15000, 10000, @ pi/4
Estimated defocus values        : 15002.54 , 10001.31 Angstroms
Estimated azimuth of astigmatism: 45.00 degrees
```

--------

Our internal `CtfEstimator` does okay with the radial case. The general case is still not ready for showtime imo.

```
{'radial_ctf_image.mrc': {'amplitude_contrast': 0.07,
                  'cs': 2.26,
                  'defocus_ang': 0.23518214391612505,
                  'defocus_u': 10031.766455018504,
                  'defocus_v': 9936.575638901675,
                  'pixel_size': 1,
                  'voltage': 200},
  # general asym  CTF case.
 'ctf_image.mrc': {'amplitude_contrast': 0.07,
                   'cs': 2.26,
                   'defocus_ang': 0.8767559917666062,
                   'defocus_u': 14872.00498806156,
                   'defocus_v': 14256.746881805191,
                   'pixel_size': 1,
                   'voltage': 200}}
```

I have not compared with the original Ayelet code, but we can do that at an opportune time just to rule out if this was developer error along the way. I also did not spend a ton of time trying to tweak either application to get better results. Given a doctored input, my expectation is it should do-the-thing out of the box.

The code for test image above is part of the proposed tutorial.

**Comment by garrettwrong** [comment]
Last time I discussed with Amit he mentioned pausing on this until catching up with Ayelet.
